setup/install_gatling/30_classification/90_clamps.rb

```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+# Clamps is a module based stress testing tool. It will generate a number of
+# users on a system to simulate agents. It will then randomize their puppet
+# agent -t times to simulate a large number of agents.
+def add_clamps_group_ca(classifier, pe_infra)
+  # The Clamps::master class manages auth.conf for the agents - by setting it
+  # to allow all.  This is insecure and worth looking at overriding at some
+  # point, but this should only be used for metrics / load testing on a private
+  # network, so shouldn't be an issue short term.
+  clamps_ca_group = {
+    'name'    => 'Clamps CA',
+    'rule'    => ['or', ['=', 'name', master.node_name]], # pinned node
+    'parent'  => pe_infra['id'],
+    'classes' => {
+      'clamps::master' => {}
+    }
+  }
+  classifier.find_or_create_node_group_model(clamps_ca_group)
+end
+
+# Creates a node group for creating the fake agents. This node group should
+# match all users that are running as root.
+def add_clamps_group_agent(classifier, pe_infra)
+  clamps_agents_group = {
+    'name'    => 'Clamps - Agent Nodes',
+    'rule'    => ['and', ['=', %w[fact id], 'root'], ['=', 'name', metric.node_name]],
+    'parent'  => pe_infra['id'],
+    'classes' => {
+      'clamps::agent' => {
+        'amqpass'               => '',
+        'master'                => any_hosts_as?(:loadbalancer) ? 'puppet' : master.node_name,
+        'nonroot_users'         => options[:scale][:num_nonroot_users]       || 2,
+        'daemonize'             => options[:scale][:daemonize]               || false,
+        'mco_daemon'            => options[:scale][:mco_daemon],
+        'num_facts_per_agent'   => options[:scale][:facts_per_agent]         || 500,
+        'percent_changed_facts' => options[:scale][:percent_facts_to_change] || 15,
+        'splay'                 => options[:scale][:splay]                   || false,
+        'splaylimit'            => options[:scale][:splaylimit]              || false
+      }.reject { |_k, v| v.nil? }
+    }
+  }
+  classifier.find_or_create_node_group_model(clamps_agents_group)
+end
+
+def add_clamps_group_users(classifier, pe_infra)
+  clamps_users_group = {
+    'name'    => 'Clamps - Agent Users (non root)',
+    'rule'    => ['and', ['not', ['=', %w[fact id], 'root']]],
```

Comment:
If I read this right... it would apply to any non-root user on any node.  Where as the 'Clamps - Agent Nodes' group applies to the root user on only the metrics node.  Seems like they should both limit to the metrics node as far as I can tell.

Reply:
No, the users group gets applied to all of the generated agents (e.g `user-1-ip-10-227.0.222-amz-dev.puppet.net`). So, this rule classifies all of the non-root agents running on generated agents.